x/twap/logic_test.go

```diff
@@ -43,130 +43,278 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
-		spotPriceResult1 twapmock.SpotPriceResult
-		expRecord        types.TwapRecord
+		recordsToPreSet     types.TwapRecord
+		testPoolId          uint64
+		testDenom0          string
+		testDenom1          string
+		testTime            time.Time
+		expectedAccumulator sdk.Dec
+		expectedErr         error
 	}{
-		"0 accum start, sp change": {
-			record:           zeroAccumNoErrSp10Record,
-			spotPriceResult0: spotPriceResOne,
-			spotPriceResult1: spotPriceResOne,
-			expRecord:        sp10OneTimeUnitAccumRecord,
-		},
-		"0 accum start, sp0 err at update": {
-			record:           zeroAccumNoErrSp10Record,
-			spotPriceResult0: spotPriceResOneErr,
-			spotPriceResult1: spotPriceResOne,
-			expRecord:        withLastErrTime(sp10OneTimeUnitAccumRecord, updateTime),
+		"same time with existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime,
 		},
-		"0 accum start, sp0 err at update with nil dec": {
-			record:           zeroAccumNoErrSp10Record,
-			spotPriceResult0: spotPriceResOneErrNilDec,
-			spotPriceResult1: spotPriceResOne,
-			expRecord:        withSp0(withLastErrTime(sp10OneTimeUnitAccumRecord, updateTime), sdk.ZeroDec()),
+		"call 1 second after existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(time.Second),
+			// 1(spot price) * 1000(one sec in milli-seconds)
+			expectedAccumulator: baseRecord.P0ArithmeticTwapAccumulator.Add(sdk.NewDec(1000)),
 		},
-		"0 accum start, sp1 err at update with nil dec": {
-			record:           zeroAccumNoErrSp10Record,
-			spotPriceResult0: spotPriceResOne,
-			spotPriceResult1: spotPriceResOneErrNilDec,
-			expRecord:        withSp1(withLastErrTime(sp10OneTimeUnitAccumRecord, updateTime), sdk.ZeroDec()),
+		"call 1 second before existing record": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset0Denom,
+			testDenom1:      baseRecord.Asset1Denom,
+			testTime:        baseTime.Add(-time.Second),
+			expectedErr: fmt.Errorf("looking for a time thats too old, not in the historical index. "+
+				" Try storing the accumulator value. (requested time %s)", baseTime.Add(-time.Second)),
 		},
-		"startRecord err time preserved": {
-			record:           withLastErrTime(zeroAccumNoErrSp10Record, baseTimeMinusOne),
-			spotPriceResult0: spotPriceResOne,
-			spotPriceResult1: spotPriceResOne,
-			expRecord:        withLastErrTime(sp10OneTimeUnitAccumRecord, baseTimeMinusOne),
+		"on lexicographical order denom parameters": {
+			recordsToPreSet: baseRecord,
+			testPoolId:      baseRecord.PoolId,
+			testDenom0:      baseRecord.Asset1Denom,
+			testDenom1:      baseRecord.Asset0Denom,
+			testTime:        baseTime,
 		},
```

Comment:
This one should be testDenom0 matches baseRecord.Asset0Denom, then same for testDenom1